README.md
# backup-functions

[![Build Status](https://drone.osshelp.ru/api/badges/OSSHelp/backup-functions/status.svg)](https://drone.osshelp.ru/OSSHelp/backup-functions)

## About

This library is used for backup purposes.

Supported software:

* MySQL
* PostgreSQL
* MongoDB
* Redis
* Elasticsearch
* Clickhouse
* RabbitMQ
* SQLite
* Consul
* Gitab
* Files and directories

Upload to a remote storage:

* [AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-install.html)
* [Rclone](https://rclone.org/docs/)
* [lftp](https://lftp.yar.ru/lftp-man.html)
* [rsync](https://linux.die.net/man/1/rsync)

So, you can use the following protocols:

* FTP/SFTP/LFTP
* rsync
* AWS S3 (or any S3-compatible)
* OpenStack Swift
* Backblaze B2
* DigitalOcean Spaces
* Dropbox
* Google Cloud Storage
* Google Drive
* Microsoft Azure Blob Storage
* Microsoft OneDrive
* WebDAV
* Yandex Disk

It also supports:

* Compressing (zip, bzip2, pbzip2)
* Encryption
* Files splitting

### How to use it

> **_NOTE:_**  If you want to send backup metrics to Pushgateway in version 4.x.x, you need to install [pushgateway-functions](https://gitea.osshelp.ru/OSSHelp/pushgateway-functions).

The library must be included in the backup scripts. In the examples folder you can find examples based on `backup-functions/custom.backup` template.

How to use the library:

1. Place backup-functions.sh library in the `/usr/local/include/osshelp/` path
1. Make a backup script from the template `/usr/local/sbin/custom.backup` or from examples (or use your own ideas)
1. Run your script and check the results of it works (are there any errors?)
1. Add script to Cron job with needed schedule
1. Make sure that the script works as you expected by schedule

There’re install/update scripts in the repository. Command for installation (backup-function.sh, custom.backup template and logrotate config):

```shell
curl -s https://oss.help/scripts/backup/backup-functions/install.sh | bash
```

Or you can use this [Ansible role](https://github.com/OSSHelp/ansible-backup-functions) to install it.

If you need, you can run the custom.backup template by hands with following keys:

* `-b` or `--backup` -- runs only `make_backup` function
* `-u` or `--upload` -- runs only `upload_backup` function

This is functionality of main function. If you run custom.backup without keys it runs both functions (make_backup and upload_backup).

## FAQ

### Metrics were not sent to pushgateway. What is it and how to disable it?

By default this library tries to send metrics to local [Pushgateway](https://github.com/prometheus/pushgateway) service. If you want to disable Pushgateway usage, you can add `no_pushgateway=1` to the options section in the script.

### Default backup scheme and how can I change it?

Default backup scheme:

* 1 local copy in the `/backup` folder
* 7 daily copies in the storage
* 4 weekly copies in the storage
* 3 monthly copies in the storage

You can override it by this variables in the script:

* local_days -- how many days local copy must be kept
* remote_backups_daily -- how many daily copies you need in storage
* remote_backups_weekly -- how many weekly copies you need in storage
* remote_backups_monthly -- how many monthly copies you need in storage

### In the local backup folder one more copy than I setted. What's wrong?

That happened because of local copies being cleaned before making a new backup. So if you set the variable `local days=2` you can see 2 copies remain after cleaning + one new copy (3 local copies total).

## Author

OSSHelp Team, see <https://oss.help>


CHANGELOG.md
# Changelog

## v4.1.1

* Updated mongo_dump_all function. Now it can use both shells: new mongosh or old mongo shell. Functions works with MongoDB 3.2+. Use mongosh for MongoDB 5.0+.
* Added mongo_dump_db function. Function works with MongoDB 3.2+.

## v4.0.6

* Added managed PostgreSQL servers support (POSTGRES_SKIP_GLOBALS env variable).

## v4.0.5

* Fixed domain detection for Backblaze
* Fixed and improved domain detection for minio/awscli
* Fixed protocol detection in rclone_sync()
* Improved domain detection in upload functions (minio/awscli/rclone)

## v4.0.0

* Metric export completely switched to pushgateway-functions
* Removed no_rmt_bckp_chck flag
* Minio-client, awscli and rclone upload functions received optional argument to disable uploaded files checking
* Fixed backup_required_space_bytes metric value generation
* Various syntax adjustments
* Fixed awscli profiles bug

## v3.31.3

* Adjusted metrics export in backup-only and upload-only scenarios.

## v3.31.2

* Proper error handling added to pushgateway_send_result().

## v3.31.1

* Fixed metrics export in upload_only and backup_only scenarios (was failing due to no values)

## v3.31.0

* Additional labels added to upload metrics (upload_util, upload_protocol, upload_domain). Auto-detection works only for awscli/rclone/minio-client for now.

## v3.30.3

* Fixed issues in the rabbitmq_backup function (files size and quantity counting).
* Updated global variables for remote backup size and files count.

## v3.30.2

* Fixed issues in the rclone_sync function (new rclone size format output in the version 1.57+).

## v3.30.1

* Added backup size counting in the remote storage for awscli_sync, minio_mirror and rclone_sync functions.

## v3.29.2

* Fixed variable name (backup_is_executing) in pushgateway_send_backup_start and pushgateway_send_backup_end functions.

## v3.29.1

* Fixed pushgateway_send_script_start function name.

## v3.29.0

* Added minio_mirror and minio_clean functions.

## v3.28.0

* Added awscli_purge_opts array and profiles usage to aws functions.

## v3.27.0

* Added backup and upload duration metrics.

## v3.26.1

* Adjusted curl calls for elasticsearch functions (missing header).

## v3.26.0

* Added function for counting files size and quantity.

## v3.25

* Fixed format parsing error (Pushgateway).

## v3.24

* Updated Pushgateway URL (resolved conflicts between scripts with the same name in difrent containers on the server).

## v3.23

* Updated Pushgateway function
* Fixed error handling

## v3.22

* Adjusted metrics generation for pushgateway.

## v3.21

* Renamed script_failure to backup_script_failure for consistency.

## v3.20

* Updated Pushgateway function
* Bugs fixed

## v3.19

* Added pushgateway functions

## v3.18

* Added rabbitmq_backup function. It installs rabbitmqadmin if necessary, exports RabbitMQ configuration, and compresses RabbitMQ datadir.
* Fixed password extraction from /root/.my.cnf for mysql_xtra_backup_db and mysql_xtra_backup_all functions

## v3.17

* Added a new version of mongo_dump_all function. The old version was renamed to mongo_dump_all_old. The new version works faster and requires less disk space and disk utilization for the dumping process.

## v3.16

* Added mysql_xtra_backup_opts array to the mysql_xtra_backup_db and mysql_xtra_backup_all functions

## v3.15

Updated rclone_purge function (added rclone_purge_opts array)
Updated excludes for shellcheck

## v3.14

* Added mysql_lxc_hotcopy_all function
* Updated rclone_sync function
* Fixed shellcheck issues

## v3.13

* Updated pg_dump_all function

## v3.12

* Deleted deprecated functions for: s3cmd, swift and supload
* Added pg_repl_ctl function (set/remove replication pause for PostgreSQL)
* Updated PostgreSQL functions (added pg_repl_ctl usage)

## v3.11

* Added new functions inc_compress_dir, consul_backup and rsync_dir
* Marked as deprecated functions for s3cmd, swift and supload utilities
* Remade detect_type function
* Updated Rclone functions

## v3.10

* Added default backup scheme to the library: 1 local backup, 7 daily backups in storage, 4 weekly backups in storage, 3 monthly backups in storage
* Added clickhouse_dump_all function
* Fixed bug in selectel_upload function
* Remade archiver choosing mechanism, added pbzip2 threads autoscaling
* Fixed bug because of which number of pbzip2 threads can't be setted
* Updated custom.backup template

## v3.06

* Added opportunity to disable archiving for MySQL dumps
* Added make_flock, detect_type and main functions
* Added custom.backup template
* Updated installer

## v3.05

* Added functions for rdiff-backup (increment backups)
* Added functions for elasticsearch backup
* Improved mongo_dump_all function (the excludes of DB list)

## v3.04

* A lot of variables changed to arrays (tar_opts e.g.)
* Fixed ShellCheck issues
* Added gitlab_backup function
* Fixed awscli_sync, ftp_clean_dir and mysql_xtra_backup_* functions


backup-functions/update.sh
#!/bin/bash
# shellcheck disable=SC2015

umask 0077
export LANG=C
export LC_ALL=C

bfver=4.1.1

script_name="backup-functions.sh"
script_path="/usr/local/include/osshelp"
logrotate_config_path='/etc/logrotate.d/custom'
index_url="${REMOTE_URI:-https://oss.help/scripts/backup/backup-functions/.list}"
list_name=$(basename "${index_url}")
tmpdir="${TEMP:=/tmp}"
shacmd=$(command -v sha256sum || command -v gsha256sum 2>/dev/null)
silent=no
err=0

function show_notice() { test "${silent}" != "yes" && echo -e "[NOTICE] ${*}"; return 0; }
function show_error() { echo -e "[ERROR] ${*}" >&2;  err=1; return 1; }
function fetch_files() {
  cd "${1}" && {
    {
      wget -q -P "${1}" "${index_url}" && \
        wget -q -i "${1}/${list_name}" -P "${1}"
    } && {
      "${shacmd}" -c --status SHA256SUMS 2> /dev/null || {
        show_error "Something went wrong, checksums of downloaded files mismatch."
        "${shacmd}" -c "${1}/SHA256SUMS"
        return 1
      }
    }
  }
}
function install_files() {
  test -d "${script_path}" || mkdir -p "${script_path}"
  cd "${script_path}" && \
    mv -f "${tmp_dir}/${script_name}" "${script_path}/${script_name}" && chmod 640 "${script_path}/${script_name}" && \
      mv -f "${tmp_dir}/logrotate.conf" "${logrotate_config_path}"
}

uid=$(id -u)
test "${uid}"  != 0 && { show_error "Sorry, but you must run this script as root."; exit 1; }

tmp_dir="${tmpdir}/backup-functions.${$}"
mkdir -p "${tmp_dir}" && \
  fetch_files "${tmp_dir}" && \
    install_files "${tmp_dir}" && {
      show_notice "Library backup-functions (v${bfver}) was updated."
      show_notice "Logrotate config ${logrotate_config_path} was updated."
    }

test -d "${tmp_dir}" && rm -rf "${tmp_dir}"
test "${err}" -eq 1 && { show_error "Installation failed."; }
exit "${err}"


examples/postgres_w_upload.sh
#!/bin/bash
# shellcheck disable=SC2059,SC1091,SC2034,SC2154
. /usr/local/include/osshelp/backup-functions.sh

cbver=1.01
# general options
backup_dir='/mnt/backup'
rclone_storage='selectel:storage'
local_days=2
remote_backups_daily=''
remote_backups_weekly=12
remote_backups_monthly=12
# functions options, excludes, etc

# main functions
function make_backup() {
   show_notice "Backup process started..."
    test -d "${backup_dir}" && clean_dir "${backup_dir}" "${local_days}"

    check_free_space "${backup_dir}" && {
        pg_dump_all "${backup_dir}/${current_date:?}"
        save_backup_size "${backup_dir}/${current_date}"
    }
}

function upload_backup() {
    show_notice "Upload process started..."
    rclone_sync "${backup_dir}/${current_date}" "${rclone_storage}/${type}/${current_date}"
    rclone_purge "${rclone_storage}/${type}" "${remote_backups}"
    show_notice "Upload process ended."
}

main "${@}"


examples/mysql_and_folders_w_upload.sh
#!/bin/bash
# shellcheck disable=SC2059,SC1091,SC2034,SC2154,SC2010
. /usr/local/include/osshelp/backup-functions.sh

cbver=1.01
# general options
rclone_storage='selectel:storage'
# functions options, excludes, etc

# main functions
function make_backup() {
   show_notice "Backup process started..."
    test -d "${backup_dir}" && clean_dir "${backup_dir}" "${local_days}"

    check_free_space "${backup_dir}" && {
        mysql_dump_all "${backup_dir}/${current_date:?}/db"

        for cur_dir in $(ls -1 /home | grep -vE "${excluded_dirs}"); do
          test -d "/home/${cur_dir}" || continue
          compress_dir "/home/${cur_dir}" "${backup_dir}/${current_date}/home_${cur_dir##*/}.tar.${compress_ext:?}"
        done

        save_backup_size "${backup_dir}/${current_date}"
    }
}

function upload_backup() {
    show_notice "Upload process started..."
    rclone_sync "${backup_dir}/${current_date}" "${rclone_storage}/${type}/${current_date}"
    rclone_purge "${rclone_storage}/${type}" "${remote_backups}"
    show_notice "Upload process ended."
}

main "${@}"


backup-functions/install.sh
#!/bin/bash
# shellcheck disable=SC2015

umask 0077
export LANG=C
export LC_ALL=C

bfver=4.1.1
cbver=1.02.0

index_url="${REMOTE_URI:-https://oss.help/scripts/backup/backup-functions/.list}"
list_name=$(basename "${index_url}")
lib_name="backup-functions.sh"
lib_dir="/usr/local/include/osshelp"
default_backup_dir='/backup'
template_name='custom.backup'
script_dir='/usr/local/sbin'
logrotate_cfg_path='/etc/logrotate.d/custom'
tmpdir="${TEMP:=/tmp}"

library_was_installed=0
template_was_installed=0
logrotate_cfg_was_installed=0
default_backup_dir_was_created=0

shacmd=$(command -v sha256sum || command -v gsha256sum 2>/dev/null)
silent=no
err=0

function show_notice() { test "${silent}" != "yes" && echo -e "[NOTICE] ${*}"; return 0; }
function show_error() { echo -e "[ERROR] ${*}" >&2;  err=1; return 1; }
function logrotate_config_is_needed() { local cfg_exists=1; test -r "${logrotate_cfg_path}" || cfg_exists=0; return "${cfg_exists}"; }
function default_dir_is_needed() { local dir_exists=1; test -d "${default_backup_dir}" || dir_exists=0; return "${dir_exists}"; }
function template_installation_is_needed() {
  local script_exists=1
  script_exists=$(find "${script_dir}" -name "custom.*backup*" -type f | grep -cv mkconf)
  return "${script_exists}"
}
function fetch_files() {
  cd "${1}" && {
    {
      wget -q -P "${1}" "${index_url}" && \
        wget -q -i "${1}/${list_name}" -P "${1}"
    } && {
      "${shacmd}" -c --status SHA256SUMS 2> /dev/null || {
        show_error "Something went wrong, checksums of downloaded files mismatch."
        "${shacmd}" -c "${1}/SHA256SUMS"
        return 1
      }
    }
  }
}
function install_files() {
  test -d "${lib_dir}" || mkdir -p "${lib_dir}"
  default_dir_is_needed && {
    mkdir -p "${default_backup_dir}" && \
      chmod 700 "${default_backup_dir}" && \
        default_backup_dir_was_created=1
  }
  logrotate_config_is_needed && {
    mv "${tmp_dir}/logrotate.conf" "${logrotate_cfg_path}" && \
      logrotate_cfg_was_installed=1
  }
  template_installation_is_needed && {
    test -x "${script_dir}/${template_name}" || {
      mv "${tmp_dir}/${template_name}" "${script_dir}/${template_name}" && \
        chmod 700 "${script_dir}/${template_name}" && \
          template_was_installed=1
    }
  }
  mv -f "${tmp_dir}/${lib_name}" "${lib_dir}/${lib_name}" && \
    chmod 640 "${lib_dir}/${lib_name}" && \
      library_was_installed=1
}
function show_notices() {
  test "${default_backup_dir_was_created}" -eq 1 && \
    show_notice "Directory ${default_backup_dir} was created."
  test "${logrotate_cfg_was_installed}" -eq 1 && \
    show_notice "Logrotate config was installed as ${logrotate_cfg_path}."
  test "${template_was_installed}" -eq 1 && \
    show_notice "Backup script template (v${cbver}) was installed as ${script_dir}/${template_name}. Make sure it is set in order before adding a cron job!"
  test "${library_was_installed}" -eq 1 && {
    show_notice "Library ${lib_name} (v${bfver}) was installed as ${lib_dir}/${lib_name}."
    show_notice "Use \"osscli check backup\" when backup script is done. Check https://oss.help/kb1453 for additional information."
  }
}

uid=$(id -u)
test "${uid}" != 0 && { show_error "Sorry, but you must run this script as root."; exit 1; }

tmp_dir="${tmpdir}/backup-functions.${$}"
mkdir -p "${tmp_dir}" && \
  fetch_files "${tmp_dir}" && \
    install_files "${tmp_dir}" && \
      show_notices

test -d "${tmp_dir}" && rm -rf "${tmp_dir}"
test "${err}" -eq 1 && { show_error "Installation failed."; }
exit "${err}"
